internal/controller/object/syncer.go

```diff
@@ -77,6 +81,16 @@ func (s *SSAResourceSyncer) GetObservedState(_ context.Context, obj *v1alpha2.Ob
 // server-side apply on the object's manifest to see what the object would look
 // like if it were applied and extracting the managed fields from that.
 func (s *SSAResourceSyncer) GetDesiredState(ctx context.Context, obj *v1alpha2.Object, manifest *unstructured.Unstructured) (*unstructured.Unstructured, error) {
+	cachedDesired, cachedHash := s.desiredStateCache.GetState()
+	// Note(erhancagirici): cache assumes the raw manifest is the sole factor
+	// affecting the desired state of the upstream k8s object.
+	// Any further development in the v1alpha2.Object semantics
+	// affecting the desired state, should include it in the hash.
+	manifestSum := sha256.Sum256(obj.Spec.ForProvider.Manifest.Raw)
+	manifestHash := hex.EncodeToString(manifestSum[:])
+	if cachedDesired != nil && cachedHash == manifestHash {
```

Comment:
Could we hide this logic in the DesiredStateCache implementation? Something like:

GetStateFor(obj)
SetStateFor(obj, state)